### PR TITLE
add completion rules for the tox tool

### PR DESCRIPTION
**Pull request by gaborbernat**
documentation for it at https://tox.readthedocs.io/en/latest/

### CODE REVIEW

File: completions/tox [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+{
+    local cur prev words cword
+    _init_completion || return
+
+    case $prev in
+        -h|--help|--version)
+            return
+            ;;
+        -e)
+            local envs=`tox --listenvs-all`
```
Comment (scop) [inline completions/tox]: Use `$(...)` instead of backticks. Also, `tox` should be invoked as `$1` here.

File: completions/tox [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+            COMPREPLY=( $(compgen -W "${envs}" -- ${cur}) )
```
Comment (scop) [inline completions/tox]: Add spaces around `$( ... )`, and quote $cur, and remove superfluous curly braces from envs and cur.

File: completions/tox [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+            return 0
```
Comment (scop) [inline completions/tox]: Just return, no need for 0.

File: completions/tox [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+            ;;
+    esac
```
Comment (scop) [inline completions/tox]: Bunch of options are lacking completions, e.g. -c --installpkg, -i, --result-json etc etc. At least, ones that take an argument should have a non-completion installed (like in the --help etc case) if their arguments are not completed.

Comment (gaborbernat) [reply completions/tox]: @scop I thought the parse help  will take care of that, so do I need to enlist the no completion for all that don't take an argument?

**Comment by gaborbernat** [comment]
@scop is now ok? Thanks!

File: completions/tox (hunk 4 above)
Comment (scop) [reply completions/tox]: No, help parsing doesn't do anything about option arguments. At least -c, --installpkg, -i, --result-json, --hashseed, --force-dep, and --workdir require an argument, so the completion should either provide argument completions for those as appropriate (like what is currently done for -e), or provide a non-completion for them, which is usually handled by just adding those options to the same case where --help and friends are.

File: completions/tox [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+{
+    local cur prev words cword
+    _init_completion || return
+
+    case $prev in
+        -h|--help|--version)
+            return
+            ;;
+        -e)
+            local envs=$( $1 --listenvs-all )
```
Comment (scop) [inline completions/tox]: $1 should be enclosed in quotes.

**Comment by obestwalter** [comment]
@gaborbernat - would be great if that would be fixed up and @scop can merge - looks like there is only a var enclosure missing ... 🥇

**Comment by gaborbernat** [comment]
@obestwalter will do today

**Comment by gaborbernat** [comment]
@scop done

**Comment by scop** [comment]
Looks like one of my review comments was lost in the "outdated" sections, so reposting:

No, help parsing doesn't do anything about option arguments. At least -c, --installpkg, -i, --result-json, --hashseed, --force-dep, and --workdir require an argument, so the completion should either provide argument completions for those as appropriate (like what is currently done for -e), or provide a non-completion for them, which is usually handled by just adding those options to the same case where --help and friends are.

**Comment by gaborbernat** [comment]
@scop sorry for missing that, fixed now it should be complete; although for some reason locally the ini file completion doesn't seem to work; any ideas? thanks!

**Comment by scop** [comment]
Merged with some code style tweaks, will push a fix for an error message I got for -e separately. ini completion works fine here.

**Comment by gaborbernat** [comment]
@scop thanks a lot!

**Pull request merged**